services/src/main/scala/cromwell/pipeline/service/GitLabProjectVersioning.scala

```diff
@@ -0,0 +1,51 @@
+package cromwell.pipeline.service
+import java.net.URLEncoder
+import java.nio.file.Path
+
+import cromwell.pipeline.datastorage.dto.{ Project, ProjectFile, Version }
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+class GitLabProjectVersioning(httpClient: HttpClient)(
+  implicit executionContext: ExecutionContext
+) extends ProjectVersioning[VersioningException] {
+
+  lazy val URL: String = "https://gitlab.com/api/v4"
+  lazy val NAMESPACE: String = "AdminLogin%2F"
+  lazy val TOKEN: Map[String, String] = Map("PRIVATE-TOKEN" -> "")
+
+  override def updateFile(project: Project, projectFile: ProjectFile): AsyncResult[String] = ???
+  override def updateFiles(project: Project, projectFiles: ProjectFiles): AsyncResult[List[String]] = ???
+
+  override def createRepository(project: Project): AsyncResult[Project] = ???
+
+  override def getFiles(project: Project, path: Path): AsyncResult[List[String]] = ???
+  override def getProjectVersions(project: Project): AsyncResult[Project] = ???
+  override def getFileVersions(project: Project, path: Path): AsyncResult[List[Version]] = ???
+  override def getFilesVersions(project: Project, path: Path): AsyncResult[List[Version]] = ???
+  override def getFileTree(project: Project, version: Option[Version]): AsyncResult[List[String]] = ???
+
+  override def getFile(project: Project, path: Path, version: Option[Version]): AsyncResult[String] = {
+    val projectId: String = project.projectId.value
+    val filePath: String = URLEncoder.encode(path.toString, "UTF-8")
+    val fileVersion: String = version.map((el) => el.value).getOrElse("master")
+
+    def responseFuture = httpClient.get(
```

Comment:
Why def? BTW, you can combine initialization with next map.

Reply:
done, thanks!)